crash on compilable code: assertion failed: Trying to access the this of another class

scalac can compile but dotc crashes.


$ cat abc.scala
```
class A { { val x = this } }
class B(x: Int) {
  class C(x: Int)
      extends B({
        val test = this
        x
      }) {
    def this() = {
      this({
        1
      })
    }
  }
}
```



$ dotc abc.scala
```
java.lang.AssertionError: assertion failed: Trying to access the this of another class: tree.symbol = class B, class symbol = class B$C compilation unit: abc.scala
	at scala.Predef$.assert(Predef.scala:165)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:344)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:291)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:90)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder$$anonfun$genNormalBlock$1$1.apply(BCodeBodyBuilder.scala:859)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder$$anonfun$genNormalBlock$1$1.apply(BCodeBodyBuilder.scala:859)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genNormalBlock$1(BCodeBodyBuilder.scala:861)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlock(BCodeBodyBuilder.scala:893)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:409)
<snipped>
```

Thanks @helloqirun , report confirmed.

Following minimized version produce the same stack trace:

```Scala
class B(x: Int) {
  class C(x: Int) extends B({val test = this; x})
}
```

Compile with `-Ycheck:all` produce following checking exception after `lambdaLift, elimStaticThis, flatten, restoreScopes`:

```
Exception in thread "main" java.lang.AssertionError: assertion failed: error while typing this, value test is not contained in class B
	at scala.Predef$.assert(Predef.scala:165)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedThis(TreeChecker.scala:358)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1455)
```

--- a/compiler/src/dotty/tools/dotc/transform/Erasure.scala
+++ b/compiler/src/dotty/tools/dotc/transform/Erasure.scala
@@ -82,7 +82,7 @@ class Erasure extends Phase with DenotTransformer { thisTransformer =>
     assertErased(tree)
     tree match {
       case res: tpd.This =>
-        assert(!ExplicitOuter.referencesOuter(ctx.owner.enclosingClass, res),
+        assert(!ExplicitOuter.referencesOuter(ctx.owner.lexicallyEnclosingClass, res),
           i"Reference to $res from ${ctx.owner.showLocated}")
       case ret: tpd.Return =>
         // checked only after erasure, as checking before erasure is complicated
@@ -389,7 +389,7 @@ object Erasure extends TypeTestsCasts{
     }
 
     override def typedThis(tree: untpd.This)(implicit ctx: Context): Tree =
-      if (tree.symbol == ctx.owner.enclosingClass || tree.symbol.isStaticOwner) promote(tree)
+      if (tree.symbol == ctx.owner.lexicallyEnclosingClass || tree.symbol.isStaticOwner) promote(tree)
       else {
         ctx.log(i"computing outer path from ${ctx.owner.ownersIterator.toList}%, % to ${tree.symbol}, encl class = ${ctx.owner.enclosingClass}")
         outer.path(tree.symbol)

--- a/compiler/src/dotty/tools/dotc/transform/ExplicitOuter.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ExplicitOuter.scala
@@ -330,7 +330,7 @@ object ExplicitOuter {
     /** The path of outer accessors that references `toCls.this` starting from
      *  the context owner's this node.
      */
-    def path(toCls: Symbol, start: Tree = This(ctx.owner.enclosingClass.asClass)): Tree = try {
+    def path(toCls: Symbol, start: Tree = This(ctx.owner.lexicallyEnclosingClass.asClass)): Tree = try {
       def loop(tree: Tree): Tree = {
         val treeCls = tree.tpe.widen.classSymbol
         val outerAccessorCtx = ctx.withPhaseNoLater(ctx.lambdaLiftPhase) // lambdalift mangles local class names, which means we cannot reliably find outer acessors anymore

--- a/tests/pos/i1756.scala
+++ b/tests/pos/i1756.scala
@@ -1,5 +1,4 @@
-/*
- * class A { { val x = this } }
+class A { { val x = this } }
 class B(x: Int) {
   class C(x: Int)
       extends B({
@@ -13,7 +12,7 @@ class B(x: Int) {
     }
   }
 }
-*/
+
 // Minimized version
 class D(x: Int) {
   class E(x: Int) extends D({val test = D.this; x})

--- a/tests/pos/i1756.scala
+++ b/tests/pos/i1756.scala
@@ -1,0 +1,21 @@
+/*
+ * class A { { val x = this } }
+class B(x: Int) {
+  class C(x: Int)
+      extends B({
+        val test = this
+        x
+      }) {
+    def this() = {
+      this({
+        1
+      })
+    }
+  }
+}
+*/
+// Minimized version
+class D(x: Int) {
+  class E(x: Int) extends D({val test = D.this; x})
+}
+

